RevitClasher/Tools.cs

```diff
@@ -167,35 +163,68 @@ public static bool belongsToView(View3D view3D, BoundingBoxXYZ bb)
             return false;
         }
 
-        public static bool getClashWithSolid(Document doc, GeometryElement a, Element b)
-        {
 
-            //ElementId id = doc.GetElement(b).Id;
-            List<ElementId> l = new List<ElementId>();
-            l.Add(b.Id);
+        /// <summary>
+        /// Detects the collision or clash between two elements. By analyzing the geometry intersection of an element (Element A) against other element (Element B)
+        /// </summary>
+        /// <param name="doc">Active revit document</param>
+        /// <param name="geometryElement">Geometry element from which the intersection will be verified (Element B)</param>
+        /// <param name="element">Element to verify the presence of intersection (Element A)</param>
+        /// <returns>A Boolean flag that indicates whether or not there is a clash between elements. True = Clash detected. False = No clash detected</returns>
+        public static bool getClashWithSolid(Document doc, GeometryElement geometryElement, Element element)
+        {
+            // Elements list, This list contains the Id's of element A
+            List<ElementId> elementsList = new List<ElementId>();
+            elementsList.Add(element.Id);
 
-            bool output = false;
+            bool clashOutputFlag = false;
 
             //Look for each solid in linked model
-            foreach (var geomObj in a)
+            foreach (GeometryObject geomObj in geometryElement)
             {
-                Solid geomSolid = geomObj as Solid;
-                if (null != geomSolid)
+                // If solid is present we perform a solid intersection filter.
+                // and later, perform a filter iteration with the element to clash (Element A vs Element B)
+                if(geomObj is Solid)
                 {
+                    Solid solidObject = geomObj as Solid;
+                    ElementIntersectsSolidFilter elementIntersectsSolidFilter = new ElementIntersectsSolidFilter(solidObject);
+                    FilteredElementCollector collector = new FilteredElementCollector(doc, elementsList).WherePasses(elementIntersectsSolidFilter);
+                    
+                    // If collector count is zero, this means that Solid intersection can't be possible with the given element (ie: Fabrication Parts).
+                    if(collector.Count() == 0)
+                    {
+                        BoundingBoxXYZ bbox = geometryElement.GetBoundingBox();
+                        Outline outline = new Outline(bbox.Min, bbox.Max);
+                        BoundingBoxIntersectsFilter bbFilter = new BoundingBoxIntersectsFilter(outline);
+                        collector = new FilteredElementCollector(doc, elementsList).WherePasses(bbFilter);
+                    }
```

Comment:
In this code, i verify it is possible to perform the Intersection with solids. If not a bounding box intersection will be performed.